Gutenboarding: Fix "Choose a domain I own" layout on mobile

StefanNieuwenhuis opened the pull request:
#### Changes proposed in this Pull Request
The copy of the "Choose a domain I own" row on the Domain Picker is unreadable on mobile devices because the CTA-link is on the same row. This PR contains a fix that stacks (i.e. placed on separate rows) both on mobile viewports.

#### Technical changes

* Update HTML semantics (i.e. from `<label/>` to `<div/>`) to align with the solution provided in the `/start` flow.
* Update flexbox alignment (mobile-first approach).
* Overrule flexbox alignment on larger-than-mobile-viewports.
* Update padding for `<ArrowButton/>`. We have to revisit this while working on #48568.

#### Expected Result
**Mobile**
<img width="381" alt="Screenshot 2021-01-25 at 14 07 32" src="https://user-images.githubusercontent.com/12104589/105717896-f17c9d00-5f20-11eb-95de-4f5ada32a6c7.png">

**Desktop**

<img width="778" alt="Screenshot 2021-01-25 at 15 22 13" src="https://user-images.githubusercontent.com/12104589/105718040-1e30b480-5f21-11eb-9dab-487ed21450c8.png">

#### Testing instructions 

#### Gutenboarding

1. Go to `/new/domains-modal`
2. Confirm that:
    * [x] Both copy & CTA-link are on the same row
    * [x] The complete row is clickable and clicking redirects you to the domain's transfer/mapping view
    * [x] Clicking the CTA-link redirects you to the domain's transfer/mapping view.
3. Open DevTools and switch to a mobile viewport (e.g. `iPhone X`).
4. Revisit `/new/domains-modal`.
5. Confirm that:
    * [x] Copy & CTA-link are stacked and readable
    * [x] The complete row is clickable and clicking redirects you to the domain's transfer/mapping view.
    * [x] Clicking the CTA-link redirects you to the domain's transfer/mapping view
6. Switch from English to a more verbose language (e.g French) and repeat steps 1-6.

#### Focused Launch
1. Go to `/page/[UNLAUNCHED_SITE_ID].wordpress.com/home?flags=create/focused-launch-flow`.
2. Click "Launch" to open the Focused Launch flow.
3. Click on "View All Domains".
4. Confirm that:
    * [x] Both copy & CTA-link are on the same row
    * [x] The complete row is clickable and clicking redirects you to the domain's transfer/mapping view.
    * [x] Clicking the CTA-link redirects you to the domain's transfer/mapping view.
5. Open DevTools and switch to a mobile viewport (e.g. `iPhone X`).
6. Revisit `/page/[UNLAUNCHED_SITE_ID].wordpress.com/home?flags=create/focused-launch-flow`.
7. Confirm that:
    * [x] Copy & CTA-link are stacked and readable
    * [x] The complete row is clickable and clicking redirects you to the domain's transfer/mapping view
    * [x] Clicking the CTA-link redirects you to the domain's transfer/mapping view.
8. Switch from English to a more verbose language (e.g French) and repeat steps 1-7.

Fixes #49171

StefanNieuwenhuis committed:
Initial commit

StefanNieuwenhuis committed:
Update HTML semantics

StefanNieuwenhuis committed:
Update flexbox alignment

StefanNieuwenhuis committed:
Update the padding for the ArrowButton component

StefanNieuwenhuis committed:
Update vertical alignment on desktop

ciampo commented on packages/onboarding/src/action-buttons/style.scss:
```diff
@@ -92,6 +92,9 @@ button.action_buttons__button.components-button {
 		color: var( --mainColor );
 
+		// @TODO: We have to revisit the ArrowButton's padding when we pick up #48568 (https://github.com/Automattic/wp-calypso/issues/48568)
+		padding-left: 0;
```
Just leaving this comment here to mark the TODO item

ciampo commented on packages/domain-picker/src/domain-picker/style.scss:
```diff
@@ -109,6 +109,17 @@ $accent-blue: #117ac9;
+		align-items: flex-start;
+
+		@include break-medium {
+			flex-direction: initial;
```
This may be more explicit:

```suggestion
			flex-direction: row;
```

Is there a reason why you'd choose `initial` over `row`?

StefanNieuwenhuis replied:
> Is there a reason why you'd choose `initial` over `row`?

Nope, it's just a matter of taste. Like we discussed here p1611667274030700-slack-GLMDTNTFD, we have to define conventions for this kind of situations. When it comes to `flexbox`, we have chosen to be more explicit to improve readability. I accepted your suggestion. Thanks 👍

ciampo commented on packages/domain-picker/src/domain-picker/use-your-domain-item.tsx:
```diff
@@ -15,7 +15,9 @@ interface Props {
 const UseYourDomainItem: React.FunctionComponent< Props > = ( { onClick } ) => {
 	return (
-		<label className="domain-picker__suggestion-item contains-link">
+		/* eslint-disable jsx-a11y/click-events-have-key-events */
+		/* eslint-disable jsx-a11y/interactive-supports-focus */
+		<div role="button" className="domain-picker__suggestion-item type-link" onClick={ onClick }>
```
This looks very hacky (and it looked hacky as well prior to this PR).

- Why are we not using a `<button />`? Is it because there's another `<button />` nested inside?
- In general, nesting interactive HTML elements within each other (e.g button inside a label, or button inside a button) goes against the HTML spec
- We may want to consider that [the rest of the domain suggestion items have a different wrapping HTML element based on a prop](https://github.com/Automattic/wp-calypso/blob/fae0931f950bd423f99496549638d9ee09e42a0a/packages/domain-picker/src/domain-picker/suggestion-item.tsx#L44-L57) — should we have this component behave the same way, so that it's always aligned with the rest of the items?

ciampo reviewed (changes requested):
Thank you for improving the layout! It looks much better on mobile now :)

I left some comments, and here's some additional feedback from my first round of review:

- Can we improve the vertical alignment of text and arrow?
<img src="https://user-images.githubusercontent.com/1083581/105837657-4e369100-5fcf-11eb-983d-91d03c550456.png" width="250" />

- Can we align the "Use a domain I own >" button to the right, similarly to the items above?

<img src="https://user-images.githubusercontent.com/1083581/105837796-7c1bd580-5fcf-11eb-8f23-360d4022180f.png" />

- While following the testing instructions, whenever I click and get redirected, the query params in the "redirected" URL are different from the testing instructions (maybe you just forgot to update them?):
  - From gutenboarding, I got redirected to `/start/domains/use-your-domain?source=http://calypso.localhost:3000/new/domains`
  - From Focused Launch, I got redirected to `https://wordpress.com/start/new-launch/domains-launch/use-your-domain?siteSlug=teststart366187286.wordpress.com&source=http://calypso.localhost:3000/page/teststart366187286.wordpress.com/home?flags=create/focused-launch-flow` — note that I got redirected to wordpress.com, even though I was testing on my local machine, and that the `source` query param contains the full URL (and not just the path)

StefanNieuwenhuis committed:
Update packages/domain-picker/src/domain-picker/style.scss

Be more explicit when it comes to `flexbox`.

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>

StefanNieuwenhuis committed:
Update button + chevron alignment

StefanNieuwenhuis commented:
> Can we improve the vertical alignment of text and arrow?

Absolutely and thanks for addressing this 👍 I've updated the component by aligning the icon with the copy's baseline w/ a negative margin since the `svg` contains whitespace (`5px`) around the chevron's path. 

> Can we align the "Use a domain I own >" button to the right, similarly to the items above?

Sure! I've updated the component.

> While following the testing instructions, whenever I click and get redirected, the query params in the "redirected" URL are different from the testing instructions (maybe you just forgot to update them?):

I've updated the testing instructions.

ciampo commented:
I started testing and noticed that on mobile the "Use a domain I own" button doesn't align correctly to the left

<img src="https://user-images.githubusercontent.com/1083581/105984441-69220780-609a-11eb-8972-a10ee8510eea.png" width="400" />

Also, what do you think about https://github.com/Automattic/wp-calypso/pull/49249#discussion_r564448003?

StefanNieuwenhuis committed:
Update ArrowButton alignment

ciampo reviewed (approved):
Code changes LGTM, tests as per instructions — although https://github.com/Automattic/wp-calypso/pull/49249/files#r564448003? It hasn't been addressed yet.

@StefanNieuwenhuis what do you think? I understand that it's outside the scope of this PR, but at the very minimum we should discuss it and/or open a separate issue about it

StefanNieuwenhuis committed:
Refactor use-a-domain-I-own semantics

The current semantics was a bit hacky and unaligned with the other domain suggestions, which are wrapped in a (surprise, surprice) wrapper component. To wrap our component, the wrapper is abstracted from the suggestion-item and moved to a seperate file.

StefanNieuwenhuis commented:
> Code changes LGTM, tests as per instructions — although https://github.com/Automattic/wp-calypso/pull/49249/files#r564448003? It hasn't been addressed yet.
> 
> @StefanNieuwenhuis what do you think? I understand that it's outside the scope of this PR, but at the very minimum we should discuss it and/or open a separate issue about it

Thanks for reviewing & approving! 
Regarding your comment, sorry, I missed it entirely! I've abstracted the wrapper into a separate file. @ciampo, could you have another look, please?

ciampo commented on packages/domain-picker/src/domain-picker/suggestion-item-wrapper.tsx:
```diff
@@ -0,0 +1,36 @@
+type WrappingComponentProps = WrappingComponentAdditionalProps &
+	( React.HTMLAttributes< HTMLDivElement > | React.ButtonHTMLAttributes< HTMLButtonElement > );
+
+const WrappingComponent = React.forwardRef< HTMLButtonElement, WrappingComponentProps >(
+	( { type, disabled, ...props }, ref ) => {
+		if ( type === 'button' ) {
+			return <div { ...( props as React.HTMLAttributes< HTMLDivElement > ) } />;
+		}
+		return (
+			<button
+				ref={ ref }
+				disabled={ disabled }
+				{ ...( props as React.ButtonHTMLAttributes< HTMLButtonElement > ) }
+			/>
+		);
+	}
+);
```
I'm still not a fan of the markup that this component outputs. Also, it's confusing that when `type === button`, the HTML element returned is a `div`.

We should open a follow-up issue to investigate the semantics of the domain picker a bit better, given all of its variations, and come up with more accessible markup/semantics.

StefanNieuwenhuis replied:
Created #49462

ciampo reviewed (approved):
LGTM! 

Feel free to merge once these two comments ([this](https://github.com/Automattic/wp-calypso/pull/49249#discussion_r566798781) and [this](https://github.com/Automattic/wp-calypso/pull/49249#discussion_r566806989)) are addressed  🚢

ciampo commented on packages/domain-picker/src/domain-picker/use-your-domain-item.tsx:
```diff
@@ -8,14 +8,22 @@ import { ArrowButton } from '@automattic/onboarding';
+		<WrappingComponent
+			type="button"
+			className="domain-picker__suggestion-item type-link"
+			ref={ useYourOwnDomainItemRef }
```
Forgot to mention — we can probably delete this line (and the `useYourOwnDomainItemRef` variable declaration)

StefanNieuwenhuis committed:
Remove unused ref
